### DISCUSSION TITLE
Android Build Cache strategies in CI

### DISCUSSION THREAD

**Discussion by tinder-inakivillar** [MobileNativeFoundation/discussions#58] [General]
In Tinder Android, we are using two layers of caching in CI:

* Build Local Cache. Storing dependencies, wrapper by job type
* Remote Cache. Storing outputs of the tasks in an S3 bucket. 

All the builds can push to the Remote cache, builds on main branch always enjoy the benefit from previously-stored task outputs making stable the duration:

<img width="1485" alt="Screen Shot 2021-03-10 at 10 32 37 AM" src="https://user-images.githubusercontent.com/55110771/110679229-1739ca80-818c-11eb-8779-27531977f896.png">

**Comment by zachgrayio** [comment]
Nice, have a few questions

- Do you know what the cache hit rate looks like?
- are developers able to access this same cache? if so, are local clean builds comparable? And what is the hit rate like for typical local builds? 
- also curious if you've tried any of the alternate cache backends that might be faster than s3, I think I saw a redis plugin somewhere that looked interesting, and we recently implemented a plugin of our own to adapt Gradle to our cache solution even

**Reply by tinder-inakivillar** [reply]
hi @zachgrayio 

PRs by Feature teams have an excellent hit ratio, around 90%. Small iterations benefit from the remote cache, reduces memory consumption.

A clean build from local environment is around 99%.  
But it's not the typical use case in a project with a high volume of changes.  When developer integrates changes from main branch in local environment,  the cache ratio decreases because he has already changes.   I would say in this scenario, around 75%.  
We try to pre-populate the cache with the results of main branch and  prs but is one field that requires more investigation. 

We are using two types of Remote Cache, S3, that works in our CI environment. And another one with Gradle Enterprise(local environment and CI). The performance of both instances is good. With Redis, I did a poc but not tested in a production environment. What's your experience with Redis?

**Comment by hilsenrat** [comment]
Thanks for sharing it, @tinder-inakivillar.  

You mentioned that you're using both Gradle Enterprise and S3 backends at your CI. Can you please advise what purpose is served by each?  
We're currently evaluating caching Gradle in Google Cloud Storage for our CI system. Do you use a specific plugin for uploading/downloading the build cache from S3? Or it's something your team is maintaining?

Thanks again for your insights.

**Reply by tinder-inakivillar** [reply]
The project using the S3 instance as Remote Cache is a CI environment isolated. The Gradle Enterprise Cache instance is used in a private network for local builds and another CI environment. 
One of the reasons of using S3 is because the CI provider is using aws and we can define same regions/zones. 

We are using a custom S3 plugin, but there are different examples in GH. In the case of GCS you have this one https://github.com/tehlers/gradle-gcs-build-cache
